### PR TITLE
Add message and platform targets

### LINKED ISSUE #2334: Add a separate CMake target for communicate
Although I refactored a bit our CMake system, there are more to be done in our CMake. I have some in mind and other people may help.

So far I'm using MPI::MPI_CXX target directly and it requires protection with `if (HAVE_MPI)`. Because our Communicate class is a layer to serve both MPI and non-MPI cases. It would be better to create a communicate target which hides both cases. Another benefit of this change is to make qmcutil lighter. qmcutil has too many things inside.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Close #2334 I mentioned earlier that qmcutils target is too fat.
Now message and platform targets are added. The overall refactoring will be months as things are gradually tackled.
I will expand Platform folder as GPU works are gradually added.

## What type(s) of changes does this code introduce?
- Build related changes

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
bora

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/Message/Communicate.cpp [hunk 1]
```diff
@@ -15,14 +15,14 @@
...
-#include <Configuration.h>
 #include "Message/Communicate.h"
-#include "Message/TagMaker.h"
 #include <iostream>
 #include <cstdio>
-#include <Platforms/sysutil.h>
-#include <Utilities/FairDivide.h>
 #include <fstream>
+#include "config.h"
+#include "Message/TagMaker.h"
```
Comment (markdewing) [inline src/Message/Communicate.cpp:23]: It looks like the TagMaker class is unused (?)  Maybe we could remove it?

Comment (ye-luo) [reply src/Message/Communicate.cpp]: Deleted.

**Comment by prckent** [comment]
Local checks:
```
[pk7@nitrogen git_QMCPACK_qmcpack]$ grep -n -i -e "(failed" -e "failed out" -e "START build_" out_unit
139:--- START build_gccnew Thu Apr 16 17:25:35 EDT 2020
1094:100% tests passed, 0 tests failed out of 37
1237:100% tests passed, 0 tests failed out of 65
1246:--- START build_gcccuda Thu Apr 16 17:28:38 EDT 2020
2318:100% tests passed, 0 tests failed out of 35
2329:--- START build_gcccuda_full Thu Apr 16 17:31:10 EDT 2020
3401:100% tests passed, 0 tests failed out of 35
3412:--- START build_gcccuda_complex Thu Apr 16 17:34:52 EDT 2020
4495:100% tests passed, 0 tests failed out of 48
4506:--- START build_gccnew_complex Thu Apr 16 17:37:38 EDT 2020
5470:100% tests passed, 0 tests failed out of 50
5626:98% tests passed, 1 tests failed out of 65
5635:	1155 - ntest_nexus_example_pwscf_relax_Ge_T (Failed)
5639:--- START build_gccnew_nompi Thu Apr 16 17:41:00 EDT 2020
6475:100% tests passed, 0 tests failed out of 25
6617:100% tests passed, 0 tests failed out of 65
6626:--- START build_gccnew_nompi_complex Thu Apr 16 17:43:15 EDT 2020
7449:100% tests passed, 0 tests failed out of 25
7591:100% tests passed, 0 tests failed out of 65
7600:--- START build_clangnew Thu Apr 16 17:45:25 EDT 2020
10619:100% tests passed, 0 tests failed out of 37
10630:--- START build_clangnew_complex Thu Apr 16 17:48:09 EDT 2020
13523:100% tests passed, 0 tests failed out of 50
13534:--- START build_clangnew_mixed Thu Apr 16 17:50:59 EDT 2020
17790:97% tests passed, 1 tests failed out of 37
17801:	 25 - deterministic-unit_test_hamiltonian (Failed)
17805:--- START build_clangnew_complex_mixed Thu Apr 16 17:53:34 EDT 2020
21971:98% tests passed, 1 tests failed out of 50
21982:	 38 - deterministic-unit_test_hamiltonian (Failed)
21986:--- START build_clangnew_aos Thu Apr 16 17:56:22 EDT 2020
25090:100% tests passed, 0 tests failed out of 35
25101:--- START build_clangnew_complex_aos Thu Apr 16 17:58:51 EDT 2020
27960:100% tests passed, 0 tests failed out of 48
```

**Comment by ye-luo** [comment]
I'm evolving things in parallel. So some simple notes was written but not on this side.

This is Platform is intended for vendor specific device codes.
For example,
math libraries, For example, BLAS/LAPACK/FFT,
runtime functions. For example, pinned memory allocation, device initialization.

Work In Progress. We will learn how to build abstraction and make things organized.
Host # CPU, OS
Accelerator # this may be an abstraction of things below.
OpenMP
OneAPI
CUDA
HIP

OutputManager is used to control our output. It is an abstraction on OS streams. So it is more like runtime functions. Utilities is clearly a bad place. Is there a better place than Platforms/Host, I don't know.

**Comment by ye-luo** [comment]
We will gradually figure out the best way as we go.
@prckent can we merge this PR and keep the discussion else where?

**Comment by prckent** [comment]
I was hoping @PDoakORNL would comment as you originally tagged him for review, but I don't know his availability.

**Comment by ye-luo** [comment]
I was expecting @PDoakORNL as the non ANL reviewer but he was busy and not around. I think he can do post merge review if he has other points to add.

**Pull request merged**